### PR TITLE
Improve font scaling for Fluent text

### PR DESCRIPTION
**Pull request by mischreiber**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

Previously, we've been scaling all text using the ramp for Apple's `body` text style. This works pretty well for _most_ text sizes, but results in too much growth for large text and too little for smaller text. This is because each text style has a fully custom, hand-built scaling ramp, which we were not respecting.

A complication: our Fluent typography system technically supports any arbitrary text size for a component, which won't neatly map to Apple's system.

Solution: let's inspect font sizes and intelligently map them to specific text styles for the purposes of size scaling.

### Verification

I added a new Typography demo controller and used it to compare before/after shots of the different typefaces:

| Size     | Before            | After               | Apple (as FYI) |
|----------|-------------------|---------------------|----------------|
| Standard ("Large") size | before-l |  after-l | apple-l |
| Largest accessibility size | before-a11y | after-a11y | apple-a11y |

### Pull request checklist

This PR has considered:
- [ ] Light and Dark appearances
- [ ] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/1169)

### CODE REVIEW

**Comment by huwilkes** [comment]
Would it be possible to get shots of the Apple fonts at these sizes? Sort of hard to tell how much closer we are with just "wrong" and "right" pictures.

File: ios/FluentUI/Core/Theme/Tokens/FontInfo.swift [hunk 1]
```diff
@@ -17,47 +17,75 @@ public struct FontInfo {
...
+        (15.0, .subheadline),
+        (13.0, .footnote),
+        (12.0, .caption),
+        (11.0, .caption2)
+    ]
+
+    public var textStyle: Font.TextStyle {
+        // Defaults to smallest supported text style for mapping, before checking if we're bigger.
+        var textStyle = Font.TextStyle.caption2
+        for tuple in Self.sizeTuples {
```
Comment (huwilkes) [inline ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:46]: Should we really be going through a for loop every time someone wants to get the textStyle from a FontInfo? Even if it's not an especially long loop.

File: ios/FluentUI/Core/Theme/Tokens/FontInfo.swift [hunk 2]
```diff
@@ -17,47 +17,75 @@ public struct FontInfo {
...
+            // Named font
+            unscaledFont = font.withWeight(uiWeight(fontInfo.weight))
         } else {
-            return .systemFont(ofSize: size, weight: uiWeight(fontInfo.weight))
+            // System font
+            unscaledFont = .systemFont(ofSize: fontInfo.size, weight: uiWeight(fontInfo.weight))
+        }
+
+        if shouldScale {
+            let fontMetrics = UIFontMetrics(forTextStyle: uiTextStyle(fontInfo.textStyle))
```
Comment (huwilkes) [inline ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:85]: I can't seem to find it in the documentation, and it probably does / seems to, but does this maintain the weight that we pass in? The documentation just sort of mentions the base font and doesn't explicitly say that it'll preserve the weight of that font.

Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:85]: Yes it does -- the scaled typography in the attached screenshots maintain their weights. Without the posted weights, those fonts would look _much_ less bold.

File: ios/FluentUI/Core/Theme/Tokens/FontInfo.swift (hunk 1 above)
Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:46]: I agree that it's a little unusual. Would you prefer an additional layer of cache? I'm not sure that there's any measurable CPU hit for this constrained loop, but I can double-check.

Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:46]: Took a trace, and it looks like we spend ~1ms of time in that method, for a workload with 1.18 seconds of CPU time (just scrolling a table view up and down, constantly regenerating cells). 

I'm leaning towards living with that, but I'm open to looking for further optimization if it feels appropriate.

Comment (huwilkes) [reply ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:46]: That seems lightweight enough that we're probably okay as is. Thanks for checking!

**Review by huwilkes — APPROVED** [review]

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/TypographyTokensDemoController.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    private var aliasTokens: AliasTokens {
+        guard let fluentTheme = self.view.window?.fluentTheme else {
+            return AliasTokens()
+        }
+        return fluentTheme.aliasTokens
+    }
+
+    private struct Constants {
+        static let cellID: String = "cellID"
```
Comment (laminesm) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/TypographyTokensDemoController.swift:49]: You should be able to use `TableViewCell.identifier` instead instead of storing this in this struct.

Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/TypographyTokensDemoController.swift:49]: Ah I technically could, but I chose not to because I'm not using the Fluent `TableViewCell` here, and I'd rather not mix and match the two.

(FWIW this class is an almost perfect copy of the `ColorTokensDemoController`, which uses the same paradigm.)

**Review by harrieshin — APPROVED** [review]

File: ios/FluentUI/Core/Theme/Tokens/FontInfo.swift (hunk 1 above)
Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:46]: Upon further thinking, I agree with @huwilkes's first point. Gonna test a simple `Dictionary` lookup cache and report back on perf impact.

Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/FontInfo.swift:46]: Okay, confirmed that there's no real benefit (the dictionary lookups end up taking roughly the same amount of time as the tiny constrained loop).

File: ios/FluentUI.Tests/FontTests.swift [hunk 4]
```diff
@@ -63,16 +29,16 @@ class FontTests: XCTestCase {
         let font = UIFont.fluent(fontInfo)
         let otherFontDescriptor = UIFontDescriptor(name: name.appending("-Semibold"), size: size)
         let otherFont = UIFont(descriptor: otherFontDescriptor, size: size)
-        XCTAssertEqual(font, otherFont)
+        XCTAssertEqual(font.fontDescriptor, otherFont.fontDescriptor)
```
Comment (edjamesmsft) [inline ios/FluentUI.Tests/FontTests.swift:32]: Why did you change this check?

Comment (mischreiber) [reply ios/FluentUI.Tests/FontTests.swift:32]: Unfortunately the new way we construct fonts fails Apple's `isEqual` check, despite the fact that every observable external attribute of the font (including its visual appearance!) matches. Since their implementation of equality is opaque to us, the only reasonable workaround I could find was to check the font descriptors, which are simple enumerations of the properties of the font (e.g. weight, size, typeface name).
